src/cargo/ops/cargo_clean.rs

```diff
@@ -322,26 +327,48 @@ impl<'cfg> CleanContext<'cfg> {
             }
         };
 
-        self.config
-            .shell()
-            .verbose(|shell| shell.status("Removing", path.display()))?;
+        if self.dry_run {
+            // Concise because if in verbose mode, the path will be written in
+            // the loop below.
+            self.config
+                .shell()
+                .concise(|shell| Ok(writeln!(shell.out(), "{}", path.display())?))?;
+        } else {
+            self.config
+                .shell()
+                .verbose(|shell| shell.status("Removing", path.display()))?;
+        }
```

Comment:
Is there a reason `status` isn't being used?  Feels odd to dump paths to the screen without anything else

Reply:
The intent with `--dry-run` is to print a list of files that it would delete (that's all it does). I also didn't think there was a particular status prefix that would fit in this situation. I also presume that most paths are going to be long, and adding to that could make it harder to see. But I'm not completely opposed if there are some suggestions on what to print.

Reply:
I feel like we should be consistent in telling people what we are removing even if we then have additional dry-run specific messages listing all of the details of what was removed.

Reply:
Or to put it another way, maybe another enhancement is for the `status` to not report a path but report a description and then when we extend this to GC, it can be "Removing .crate files", "Removing git repos", etc.

Reply:
I like the idea of displaying high-level summaries for the different categories of things that could be deleted. I can maybe include something like that when rebasing this in #12634. However, the particulars seem a little complicated.

I changed the behavior here to not display paths with `--dry-run` without `--verbose`, since the individual files aren't usually particularly interesting. I did keep it with `--verbose` since I think it can potentially be useful, and is also useful for testing to see what it is doing. I was also wondering if maybe it would make sense to have the non-dry-run behave the same (that is, `cargo clean --verbose` should display all paths, not just the ones passed to rm_rf). In concept, behaving the same as Unix `rm -rv`. WDYT?

So, for example, show categories without `--verbose`, and show specific paths with `--verbose`. 

I'm debating with myself of whether or not to show "Removed xyz" after it is removed, or "Removing xyz" before it starts removing. With "Removed", it could display a file/byte summary, which could be interesting to the user ("oh wow, it just deleted 5GiB in git repos!"). With displaying "Removing" before it starts, that can provide useful information in case it fails at some point (though maybe the error message could be updated to explain which category it was trying to delete?).

Some sketches on different ideas:

non-verbose:
```
Removed target directory, 19 files, 19.6KiB
Removed 19 files, 19.7KiB total
```

(Maybe the final summary could be elided if there is only one category of thing that is removed? So the above would just be one line instead of repeating sizes twice.)

Or with `cargo clean -p foo`:
```
Removed target artifacts for package `foo`, 19 files, 19.7KiB
Removed 19 files, 19.7KiB total
```

non-verbose when garbage collecting:
```
Removed stale src caches, 23 files, 1.8MiB
Removed stale .crate caches, 4 files, 192KiB
Removed stale git repos, 192 files, 5.9MiB
Removed stale target artifacts, 36 files, 1.1GiB
Removed 255 files, 1.5GiB total
```

verbose:
```
Removing stale src caches
.... bunch of paths
Removed stale src caches, 23 files, 1.8MiB
Removing target directory
/path/to/target/.rustc_info.json
/path/to/target/CACHEDIR.TAG
/path/to/target/debug/.fingerprint/bar-402c80d9e18085ec/dep-lib-bar
/path/to/target/debug/.fingerprint/bar-402c80d9e18085ec/lib-bar.json
/path/to/target/debug/.fingerprint/bar-402c80d9e18085ec/invoked.timestamp
/path/to/target/debug/.fingerprint/bar-402c80d9e18085ec/lib-bar
/path/to/target/debug/.fingerprint/bar-402c80d9e18085ec
/path/to/target/debug/.fingerprint/foo-810d0583585bc9f9/lib-foo.json
/path/to/target/debug/.fingerprint/foo-810d0583585bc9f9/dep-lib-foo
/path/to/target/debug/.fingerprint/foo-810d0583585bc9f9/lib-foo
/path/to/target/debug/.fingerprint/foo-810d0583585bc9f9/invoked.timestamp
/path/to/target/debug/.fingerprint/foo-810d0583585bc9f9
/path/to/target/debug/.fingerprint
/path/to/target/debug/incremental/foo-hg2akmqki9nq/s-goltvoprn5-rijetd-bu4tjhm078p81fl04874cly2b/query-cache.bin
/path/to/target/debug/incremental/foo-hg2akmqki9nq/s-goltvoprn5-rijetd-bu4tjhm078p81fl04874cly2b/dep-graph.bin
/path/to/target/debug/incremental/foo-hg2akmqki9nq/s-goltvoprn5-rijetd-bu4tjhm078p81fl04874cly2b/work-products.bin
/path/to/target/debug/incremental/foo-hg2akmqki9nq/s-goltvoprn5-rijetd-bu4tjhm078p81fl04874cly2b
/path/to/target/debug/incremental/foo-hg2akmqki9nq/s-goltvoprn5-rijetd.lock
/path/to/target/debug/incremental/foo-hg2akmqki9nq
/path/to/target/debug/incremental
/path/to/target/debug/.cargo-lock
/path/to/target/debug/examples
/path/to/target/debug/deps/foo-810d0583585bc9f9.d
/path/to/target/debug/deps/libfoo-810d0583585bc9f9.rmeta
/path/to/target/debug/deps/bar-402c80d9e18085ec.d
/path/to/target/debug/deps/libbar-402c80d9e18085ec.rmeta
/path/to/target/debug/deps
/path/to/target/debug/build
/path/to/target/debug
/path/to/target
Removed target directory, 19 files, 19.7KiB
Removed 42 files, 1.9MiB total
```

Reply:
To follow up on this, @epage and I chatted a little about different ideas here. I would like to add category summaries at the end when it is finished, though those aren't very meaningful without #12634. After reviewing the current behavior, I think I would like to move forward with the behavior in this PR as-is. I'm not entirely sure how the API might change to be able to track the different categories, and it's a little challenging to do that in isolation in this PR. Is there any objection to proceeding with the behavior here? And then perhaps following up on that later?

I'm still a little undecided on the `--verbose` behavior. Currently it is a little inconsistent about what it displays. For example, a bare `cargo clean -v` just prints `Removing /path/to/target` whereas `cargo clean -p ... -v` or `cargo clean --gc` will print `Removing` for individual files. That seems like an independent issue that could be revisited at some point in the future. I don't have much of a preference either way, and it is a relatively easy change to make.